rust/connlib/shared/src/linux/etc_resolv_conf.rs

```diff
@@ -31,35 +43,35 @@ async fn configure_dns_at_paths(
     dns_config: &[IpAddr],
     resolv_path: &Path,
     backup_path: &Path,
-) -> Result<()> {
+) -> Result<(), Error> {
     if dns_config.is_empty() {
         tracing::info!("`dns_config` is empty, leaving `/etc/resolv.conf` unchanged");
         return Ok(());
     }
 
     let text = tokio::fs::read_to_string(resolv_path)
         .await
-        .map_err(Error::ReadResolvConf)?;
-    let parsed = resolv_conf::Config::parse(&text).map_err(|_| Error::ParseResolvConf)?;
+        .map_err(Error::Read)?;
+    let parsed = resolv_conf::Config::parse(&text).map_err(|_| Error::Parse)?;
 
     // Back up the original resolv.conf. If there's already a backup, don't modify it
-    match tokio::fs::File::options()
-        .write(true)
-        .create_new(true)
-        .open(backup_path)
-        .await
-    {
-        Err(error) if error.kind() == std::io::ErrorKind::AlreadyExists => {
+    // `atomicwrites` handles the fsync and rename-into-place tricks to resist file corruption
+    // if we lose power during the write.
+    let backup_file = atomicwrites::AtomicFile::new(
+        backup_path,
+        atomicwrites::OverwriteBehavior::DisallowOverwrite,
+    );
+    match backup_file.write(|f| f.write_all(text.as_bytes())) {
+        Err(atomicwrites::Error::Internal(error))
+            if error.kind() == std::io::ErrorKind::AlreadyExists =>
+        {
             tracing::info!(?backup_path, "Backup path already exists, won't overwrite");
         }
-        Err(error) => return Err(Error::WriteResolvConfBackup(error)),
-        // TODO: Would do a rename-into-place here if the contents of the file mattered more
-        Ok(mut f) => f.write_all(text.as_bytes()).await?,
+        Err(atomicwrites::Error::Internal(error)) => return Err(Error::SyncBackup(error)),
+        Err(atomicwrites::Error::User(error)) => return Err(Error::WriteBackup(error)),
+        Ok(()) => {}
     }
```

Comment:
Out of curiosity, can this be refactored to something like

```rust
let result = backup_file.write(|f| f.write_all(text.as_bytes()))

if let Err(atomicwrites::Error::Internal(result)) {
  ...
}
...
```

I've seen @thomaseizinger suggest similar modifications in the past, so just curious. Seems like it would remove the need for the `Ok(()) => {}` noop.